src/memory/AmdFlash.cc

```diff
@@ -692,11 +692,27 @@ bool AmdFlash::checkCommandProgramHelper(size_t numBytes, std::span<const uint8_
 		const Sector& sector = getSector(cmd.back().addr);
 		if (isWritable(sector)) {
 			const size_t pageMask = chip.program.fastPageSize - 1;
-			for (auto i : xrange(cmdSeq.size(), cmdSeq.size() + numBytes)) {
-				auto ramAddr = sector.writeAddress + (cmd.back().addr & ~pageMask) + (cmd[i].addr & pageMask) - sector.address;
-				uint8_t ramValue = (*ram)[ramAddr] & cmd[i].value;
-				ram->write(ramAddr, ramValue);
+			const size_t programAddress = cmd.back().addr & ~pageMask;
+			std::vector<std::optional<uint8_t>> programBuffer;
+			programBuffer.assign(chip.program.fastPageSize, std::nullopt);
+
+			// For fast program commands, if you write 0x7F and then 0xF7 to the same empty
+			// flash memory location within the same command, the final value is 0x77.
+			// Tested on: M29W640GB.
+			for (size_t i : xrange(cmdSeq.size(), cmdSeq.size() + numBytes)) {
+				auto& value = programBuffer[cmd[i].addr & pageMask];
+				value = cmd[i].value & value.value_or(0xff);
 			}
+
+			size_t ramAddr = programAddress - sector.address + sector.writeAddress;
+			for (const auto value : programBuffer) {
```

Comment:
Detail: for `optional<uint8_t>` it technically doesn't matter (the current code is also correct). But I still wonder if this is a typo or not.

My expectation for declaring the type in a range-for loop is either:
* `auto`: if you intentionally want to make a copy
* `const auto&`: if you just want to use the value (not modify it)
* `auto&`: if you want to (use and) modify it.
* (less common) an explicit type: for an intentional type-conversion

But `const auto` is a mix between the 1st and 2nd. Therefor I wonder:
* (Most likely) did you forget the `&`?
* Did you intend to make a copy? Then `const` is not wrong, but also doesn't contribute much.

Reply:
I thought I wouldn’t need `&` since it’s a “small value” that easily fits in registers, and adding the & might force it to use pointers instead. Probably am thinking too hard on this.